File: session_py/src/session_py/polyline_test.py
"""Tests for Polyline class."""

from session_py.point import Point
from session_py.polyline import Polyline
from session_py.vector import Vector


def test_polyline_new():
    points = [Point(0.0, 0.0, 0.0), Point(1.0, 0.0, 0.0), Point(0.0, 1.0, 0.0)]
    polyline = Polyline(points)
    assert len(polyline) == 3
    assert polyline.segment_count() == 2


def test_polyline_default():
    polyline = Polyline()
    assert len(polyline) == 0
    assert polyline.is_empty()
    assert polyline.segment_count() == 0


def test_polyline_length():
    points = [Point(0.0, 0.0, 0.0), Point(1.0, 0.0, 0.0), Point(1.0, 1.0, 0.0)]
    polyline = Polyline(points)
    length = polyline.length()
    assert abs(length - 2.0) < 1e-5


def test_polyline_add_point():
    polyline = Polyline([Point(0.0, 0.0, 0.0), Point(1.0, 0.0, 0.0)])
    assert len(polyline) == 2

    polyline.add_point(Point(1.0, 1.0, 0.0))
    assert len(polyline) == 3
    assert polyline.segment_count() == 2


def test_polyline_insert_point():
    polyline = Polyline([Point(0.0, 0.0, 0.0), Point(2.0, 0.0, 0.0)])

    polyline.insert_point(1, Point(1.0, 0.0, 0.0))
    assert len(polyline) == 3
    assert polyline.points[1].x == 1.0


def test_polyline_remove_point():
    points = [Point(0.0, 0.0, 0.0), Point(1.0, 0.0, 0.0), Point(2.0, 0.0, 0.0)]
    polyline = Polyline(points)

    removed = polyline.remove_point(1)
    assert removed is not None
    assert removed.x == 1.0
    assert len(polyline) == 2


def test_polyline_reverse():
    points = [Point(0.0, 0.0, 0.0), Point(1.0, 0.0, 0.0), Point(2.0, 0.0, 0.0)]
    polyline = Polyline(points)

    polyline.reverse()
    assert polyline.points[0].x == 2.0
    assert polyline.points[1].x == 1.0
    assert polyline.points[2].x == 0.0


def test_polyline_reversed():
    points = [Point(0.0, 0.0, 0.0), Point(1.0, 0.0, 0.0), Point(2.0, 0.0, 0.0)]
    polyline = Polyline(points)

    reversed_polyline = polyline.reversed()
    assert reversed_polyline.points[0].x == 2.0
    assert reversed_polyline.points[1].x == 1.0
    assert reversed_polyline.points[2].x == 0.0

    # Original should be unchanged
    assert polyline.points[0].x == 0.0


def test_polyline_add_assign_vector():
    polyline = Polyline([Point(1.0, 2.0, 3.0), Point(4.0, 5.0, 6.0)])
    v = Vector(4.0, 5.0, 6.0)
    polyline += v

    assert polyline.points[0].x == 5.0
    assert polyline.points[0].y == 7.0
    assert polyline.points[0].z == 9.0
    assert polyline.points[1].x == 8.0
    assert polyline.points[1].y == 10.0
    assert polyline.points[1].z == 12.0


def test_polyline_add_vector():
    polyline = Polyline([Point(1.0, 2.0, 3.0), Point(4.0, 5.0, 6.0)])
    v = Vector(4.0, 5.0, 6.0)
    polyline2 = polyline + v

    assert polyline2.points[0].x == 5.0
    assert polyline2.points[0].y == 7.0
    assert polyline2.points[0].z == 9.0


def test_polyline_sub_assign_vector():
    polyline = Polyline([Point(1.0, 2.0, 3.0), Point(4.0, 5.0, 6.0)])
    v = Vector(4.0, 5.0, 6.0)
    polyline -= v

    assert polyline.points[0].x == -3.0
    assert polyline.points[0].y == -3.0
    assert polyline.points[0].z == -3.0
    assert polyline.points[1].x == 0.0
    assert polyline.points[1].y == 0.0
    assert polyline.points[1].z == 0.0


def test_polyline_sub_vector():
    polyline = Polyline([Point(1.0, 2.0, 3.0), Point(4.0, 5.0, 6.0)])
    v = Vector(4.0, 5.0, 6.0)
    polyline2 = polyline - v

    assert polyline2.points[0].x == -3.0
    assert polyline2.points[0].y == -3.0
    assert polyline2.points[0].z == -3.0
    assert polyline2.points[1].x == 0.0
    assert polyline2.points[1].y == 0.0
    assert polyline2.points[1].z == 0.0


def test_polyline_display():
    polyline = Polyline([Point(0.0, 0.0, 0.0), Point(1.0, 0.0, 0.0)])
    display_str = str(polyline)
    assert "Polyline" in display_str
    assert "points=2" in display_str


def test_polyline_to_json_data():
    polyline = Polyline([Point(0.0, 0.0, 0.0), Point(1.0, 0.0, 0.0)])

    json_string = polyline.to_json_data()
    assert "Polyline" in json_string
    assert "points" in json_string


def test_polyline_from_json_data():
    polyline = Polyline([Point(1.0, 2.0, 3.0), Point(4.0, 5.0, 6.0)])

    json_string = polyline.to_json_data()
    deserialized = Polyline.from_json_data(json_string)

    assert len(deserialized) == 2
    assert deserialized.points[0].x == 1.0
    assert deserialized.points[1].x == 4.0


def test_polyline_to_json_from_json():
    points = [Point(1.0, 2.0, 3.0), Point(4.0, 5.0, 6.0), Point(7.0, 8.0, 9.0)]
    polyline = Polyline(points)

    filepath = "test_polyline.json"
    polyline.to_json(filepath)
    loaded = Polyline.from_json(filepath)

    assert len(loaded) == 3
    assert loaded.points[0].x == 1.0
    assert loaded.points[1].y == 5.0
    assert loaded.points[2].z == 9.0


def test_polyline_get_point():
    polyline = Polyline([Point(0.0, 0.0, 0.0), Point(1.0, 2.0, 3.0)])

    point = polyline.get_point(1)
    assert point is not None
    assert point.x == 1.0

    invalid = polyline.get_point(10)
    assert invalid is None

File: session_rust/src/polyline_test.rs
use crate::{Point, Polyline, Vector};

#[test]
fn test_polyline_new() {
    let points = vec![
        Point::new(0.0, 0.0, 0.0),
        Point::new(1.0, 0.0, 0.0),
        Point::new(0.0, 1.0, 0.0),
    ];
    let polyline = Polyline::new(points);
    assert_eq!(polyline.len(), 3);
    assert_eq!(polyline.segment_count(), 2);
}

#[test]
fn test_polyline_default() {
    let polyline = Polyline::default();
    assert_eq!(polyline.len(), 0);
    assert!(polyline.is_empty());
    assert_eq!(polyline.segment_count(), 0);
}

#[test]
fn test_polyline_length() {
    let points = vec![
        Point::new(0.0, 0.0, 0.0),
        Point::new(1.0, 0.0, 0.0),
        Point::new(1.0, 1.0, 0.0),
    ];
    let polyline = Polyline::new(points);
    let length = polyline.length();
    assert!((length - 2.0).abs() < 1e-5);
}

#[test]
fn test_polyline_add_point() {
    let mut polyline = Polyline::new(vec![Point::new(0.0, 0.0, 0.0), Point::new(1.0, 0.0, 0.0)]);
    assert_eq!(polyline.len(), 2);

    polyline.add_point(Point::new(1.0, 1.0, 0.0));
    assert_eq!(polyline.len(), 3);
    assert_eq!(polyline.segment_count(), 2);
}

#[test]
fn test_polyline_insert_point() {
    let mut polyline = Polyline::new(vec![Point::new(0.0, 0.0, 0.0), Point::new(2.0, 0.0, 0.0)]);

    polyline.insert_point(1, Point::new(1.0, 0.0, 0.0));
    assert_eq!(polyline.len(), 3);
    assert_eq!(polyline.points[1].x(), 1.0);
}

#[test]
fn test_polyline_remove_point() {
    let mut polyline = Polyline::new(vec![
        Point::new(0.0, 0.0, 0.0),
        Point::new(1.0, 0.0, 0.0),
        Point::new(2.0, 0.0, 0.0),
    ]);

    let removed = polyline.remove_point(1);
    assert!(removed.is_some());
    assert_eq!(removed.unwrap().x(), 1.0);
    assert_eq!(polyline.len(), 2);
}

#[test]
fn test_polyline_reverse() {
    let mut polyline = Polyline::new(vec![
        Point::new(0.0, 0.0, 0.0),
        Point::new(1.0, 0.0, 0.0),
        Point::new(2.0, 0.0, 0.0),
    ]);

    polyline.reverse();
    assert_eq!(polyline.points[0].x(), 2.0);
    assert_eq!(polyline.points[1].x(), 1.0);
    assert_eq!(polyline.points[2].x(), 0.0);
}

#[test]
fn test_polyline_reversed() {
    let polyline = Polyline::new(vec![
        Point::new(0.0, 0.0, 0.0),
        Point::new(1.0, 0.0, 0.0),
        Point::new(2.0, 0.0, 0.0),
    ]);

    let reversed = polyline.reversed();
    assert_eq!(reversed.points[0].x(), 2.0);
    assert_eq!(reversed.points[1].x(), 1.0);
    assert_eq!(reversed.points[2].x(), 0.0);

    // Original should be unchanged
    assert_eq!(polyline.points[0].x(), 0.0);
}

#[test]
fn test_polyline_add_assign_vector() {
    let mut polyline = Polyline::new(vec![Point::new(1.0, 2.0, 3.0), Point::new(4.0, 5.0, 6.0)]);
    let v = Vector::new(4.0, 5.0, 6.0);
    polyline += &v;

    assert_eq!(polyline.points[0].x(), 5.0);
    assert_eq!(polyline.points[0].y(), 7.0);
    assert_eq!(polyline.points[0].z(), 9.0);
    assert_eq!(polyline.points[1].x(), 8.0);
    assert_eq!(polyline.points[1].y(), 10.0);
    assert_eq!(polyline.points[1].z(), 12.0);
}

#[test]
fn test_polyline_add_vector() {
    let polyline = Polyline::new(vec![Point::new(1.0, 2.0, 3.0), Point::new(4.0, 5.0, 6.0)]);
    let v = Vector::new(4.0, 5.0, 6.0);
    let polyline2 = polyline + &v;

    assert_eq!(polyline2.points[0].x(), 5.0);
    assert_eq!(polyline2.points[0].y(), 7.0);
    assert_eq!(polyline2.points[0].z(), 9.0);
}

#[test]
fn test_polyline_sub_assign_vector() {
    let mut polyline = Polyline::new(vec![Point::new(1.0, 2.0, 3.0), Point::new(4.0, 5.0, 6.0)]);
    let v = Vector::new(4.0, 5.0, 6.0);
    polyline -= &v;

    assert_eq!(polyline.points[0].x(), -3.0);
    assert_eq!(polyline.points[0].y(), -3.0);
    assert_eq!(polyline.points[0].z(), -3.0);
    assert_eq!(polyline.points[1].x(), 0.0);
    assert_eq!(polyline.points[1].y(), 0.0);
    assert_eq!(polyline.points[1].z(), 0.0);
}

#[test]
fn test_polyline_sub_vector() {
    let polyline = Polyline::new(vec![Point::new(1.0, 2.0, 3.0), Point::new(4.0, 5.0, 6.0)]);
    let v = Vector::new(4.0, 5.0, 6.0);
    let polyline2 = polyline - &v;

    assert_eq!(polyline2.points[0].x(), -3.0);
    assert_eq!(polyline2.points[0].y(), -3.0);
    assert_eq!(polyline2.points[0].z(), -3.0);
    assert_eq!(polyline2.points[1].x(), 0.0);
    assert_eq!(polyline2.points[1].y(), 0.0);
    assert_eq!(polyline2.points[1].z(), 0.0);
}

#[test]
fn test_polyline_display() {
    let polyline = Polyline::new(vec![Point::new(0.0, 0.0, 0.0), Point::new(1.0, 0.0, 0.0)]);
    let display_str = format!("{polyline}");
    assert!(display_str.contains("Polyline"));
    assert!(display_str.contains("points=2"));
}

#[test]
fn test_polyline_json_serialization() {
    let polyline = Polyline::new(vec![
        Point::new(0.0, 0.0, 0.0),
        Point::new(1.0, 0.0, 0.0),
        Point::new(1.0, 1.0, 0.0),
    ]);

    let json = serde_json::to_string(&polyline).unwrap();
    let deserialized: Polyline = serde_json::from_str(&json).unwrap();

    assert_eq!(deserialized.len(), 3);
    assert_eq!(deserialized.points[0].x(), 0.0);
    assert_eq!(deserialized.points[1].x(), 1.0);
    assert_eq!(deserialized.points[2].y(), 1.0);
}

#[test]
fn test_polyline_to_json_data() {
    let polyline = Polyline::new(vec![Point::new(0.0, 0.0, 0.0), Point::new(1.0, 0.0, 0.0)]);

    let json_string = polyline.to_json_data().unwrap();
    assert!(json_string.contains("Polyline"));
    assert!(json_string.contains("points"));
}

#[test]
fn test_polyline_from_json_data() {
    let polyline = Polyline::new(vec![Point::new(1.0, 2.0, 3.0), Point::new(4.0, 5.0, 6.0)]);

    let json_string = polyline.to_json_data().unwrap();
    let deserialized = Polyline::from_json_data(&json_string).unwrap();

    assert_eq!(deserialized.len(), 2);
    assert_eq!(deserialized.points[0].x(), 1.0);
    assert_eq!(deserialized.points[1].x(), 4.0);
}

#[test]
fn test_polyline_to_json_from_json() {
    let polyline = Polyline::new(vec![
        Point::new(1.0, 2.0, 3.0),
        Point::new(4.0, 5.0, 6.0),
        Point::new(7.0, 8.0, 9.0),
    ]);

    let filepath = "test_polyline.json";
    polyline.to_json(filepath).unwrap();
    let loaded = Polyline::from_json(filepath).unwrap();

    assert_eq!(loaded.len(), 3);
    assert_eq!(loaded.points[0].x(), 1.0);
    assert_eq!(loaded.points[1].y(), 5.0);
    assert_eq!(loaded.points[2].z(), 9.0);
}

#[test]
fn test_polyline_get_point() {
    let polyline = Polyline::new(vec![Point::new(0.0, 0.0, 0.0), Point::new(1.0, 2.0, 3.0)]);

    let point = polyline.get_point(1);
    assert!(point.is_some());
    assert_eq!(point.unwrap().x(), 1.0);

    let invalid = polyline.get_point(10);
    assert!(invalid.is_none());
}

#[test]
fn test_polyline_get_point_mut() {
    let mut polyline = Polyline::new(vec![Point::new(0.0, 0.0, 0.0), Point::new(1.0, 2.0, 3.0)]);

    if let Some(point) = polyline.get_point_mut(1) {
        *point = Point::new(5.0, 6.0, 7.0);
    }

    assert_eq!(polyline.points[1].x(), 5.0);
    assert_eq!(polyline.points[1].y(), 6.0);
    assert_eq!(polyline.points[1].z(), 7.0);
}
